**NLP** **Assignment 1** (Problem Statement 33)

## Group No 86

## Group Member Names:
1.GORLA HEMA VISHAL	(2021sc04273)

2.PATTANAYAK ASIS RANJAN (2021sc04438)

3.NAVIN KUMAR	(2021sc04553)

4.ASHLESH B R	(2021sc04863)

In [2]:
import pandas as pd
import nltk
nltk.download('punkt')
nltk.download('stopwords')

# Import necessary NLP libraries
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


In [3]:
# Read the CSV file into a DataFrame
df = pd.read_csv("/content/all_kindle_review .csv")

# Display the first few rows of the DataFrame
df.head()

Unnamed: 0.1  Unnamed: 0        asin  helpful  rating  \
0             0       11539  B0033UV8HI  [8, 10]       3   
1             1        5957  B002HJV4DE   [1, 1]       5   
2             2        9146  B002ZG96I4   [0, 0]       3   
3             3        7038  B002QHWOEU   [1, 3]       3   
4             4        1776  B001A06VJ8   [0, 1]       4   

                                          reviewText   reviewTime  \
0  Jace Rankin may be short, but he's nothing to ...   09 2, 2010   
1  Great short read.  I didn't want to put it dow...   10 8, 2013   
2  I'll start by saying this is the first of four...  04 11, 2014   
3  Aggie is Angela Lansbury who carries pocketboo...   07 5, 2014   
4  I did not expect this type of book to be in li...  12 31, 2012   

       reviewerID  reviewerName                   summary  unixReviewTime  
0  A3HHXRELK8BHQG        Ridley  Entertaining But Average      1283385600  
1  A2RGNZ0TRF578I  Holly Butler   Terrific menage scenes!      1381190400  
2  A3S0H2HV6U1I7F       Merissa          Snapdragon Alley      1397174400  
3   AC4OQW3GZ919J    Cleargrace    very light murder cozy      1404518400  
4  A3C9V987IQHOQD      Rjostler                      Book      1356912000

In [4]:
#Removing punctuations, special characters and stopwords from the text column and  Converting the text to lower case.

import string

# Define function to remove punctuation and special characters
def remove_punctuation(text):
    return text.translate(str.maketrans("", "", string.punctuation))

# Define function to remove stopwords
def remove_stopwords(text):
    stop_words = set(stopwords.words("english"))
    tokens = word_tokenize(text)
    filtered_tokens = [word for word in tokens if word.lower() not in stop_words]
    return " ".join(filtered_tokens)

# Apply preprocessing to the reviewText column
df["Cleaned_Text"] = df["reviewText"].apply(remove_punctuation)
df["Cleaned_Text"] = df["reviewText"].apply(remove_stopwords)
df["Cleaned_Text"] = df["Cleaned_Text"].str.lower()

# Display the modified DataFrame
df.head()

Unnamed: 0.1  Unnamed: 0        asin  helpful  rating  \
0             0       11539  B0033UV8HI  [8, 10]       3   
1             1        5957  B002HJV4DE   [1, 1]       5   
2             2        9146  B002ZG96I4   [0, 0]       3   
3             3        7038  B002QHWOEU   [1, 3]       3   
4             4        1776  B001A06VJ8   [0, 1]       4   

                                          reviewText   reviewTime  \
0  Jace Rankin may be short, but he's nothing to ...   09 2, 2010   
1  Great short read.  I didn't want to put it dow...   10 8, 2013   
2  I'll start by saying this is the first of four...  04 11, 2014   
3  Aggie is Angela Lansbury who carries pocketboo...   07 5, 2014   
4  I did not expect this type of book to be in li...  12 31, 2012   

       reviewerID  reviewerName                   summary  unixReviewTime  \
0  A3HHXRELK8BHQG        Ridley  Entertaining But Average      1283385600   
1  A2RGNZ0TRF578I  Holly Butler   Terrific menage scenes!      1381190400   
2  A3S0H2HV6U1I7F       Merissa          Snapdragon Alley      1397174400   
3   AC4OQW3GZ919J    Cleargrace    very light murder cozy      1404518400   
4  A3C9V987IQHOQD      Rjostler                      Book      1356912000   

                                        Cleaned_Text  
0  jace rankin may short , 's nothing mess , man ...  
1  great short read . n't want put read one sitti...  
2  'll start saying first four books n't expectin...  
3  aggie angela lansbury carries pocketbooks inst...  
4  expect type book library pleased find price right

In [10]:


#Creating two objects X and y. X will be the ' Cleaned_text' which is prepared in above row, and y will be the “Rating” column.
#Vectorizing ‘Cleaned_text’ using CountVectorizer
#and TF-IDF for bigrams and display the vectors

from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer

# Create X and y objects
X = df['Cleaned_Text']
y = df['rating']

# Vectorize Cleaned_Text using CountVectorizer for bigrams
count_vectorizer = CountVectorizer(ngram_range=(2, 2))
X_count = count_vectorizer.fit_transform(X)

# Vectorize Cleaned_Text using TF-IDF for bigrams
tfidf_vectorizer = TfidfVectorizer(ngram_range=(2, 2))
X_tfidf = tfidf_vectorizer.fit_transform(X)

# Display the vectors
print("CountVectorizer (Bigrams) Vectors:")
print(X_count.toarray())

print("\nTF-IDF (Bigrams) Vectors:")
print(X_tfidf.toarray())

CountVectorizer (Bigrams) Vectors:
[[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]

TF-IDF (Bigrams) Vectors:
[[0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 ...
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]]


In [7]:
X

0        jace rankin may short , 's nothing mess , man ...
1        great short read . n't want put read one sitti...
2        'll start saying first four books n't expectin...
3        aggie angela lansbury carries pocketbooks inst...
4        expect type book library pleased find price right
                               ...                        
11995    valentine cupid vampire- jena ian another vamp...
11996    read seven books series . apocalyptic/adventur...
11997    book really n't cuppa . situation man capturin...
11998    tried use charge kindle , n't even register ch...
11999    taking instruction look often hidden world sex...
Name: Cleaned_Text, Length: 12000, dtype: object

In [15]:
#Vectorizing ‘Cleaned_Text’ using Skip-Gram and CBOW using context words sizes [1,5] and vector size as 50.

from gensim.models import Word2Vec
from nltk.tokenize import word_tokenize

from nltk.tokenize import word_tokenize

sentences = X.tolist()
# Tokenize the sentences
tokenized_sentences = [word_tokenize(sentence) for sentence in sentences]

# Define function to train Word2Vec model with different context window sizes
def train_word2vec_model(sentences, window_size, vector_size, sg):
    model = Word2Vec(sentences, sg=sg, window=window_size, vector_size=vector_size)
    return model

# Vectorize reviewText using Word2Vec with different context window sizes
context_sizes = [1, 2, 3, 4, 5]
vector_size = 50

for context_size in context_sizes:
    print(f"Word2Vec (Skip-Gram) Vectors with context window size = {context_size}:")
    model_sg = train_word2vec_model(sentences, context_size, vector_size, sg=1)

    print("Word\t\tVector")
    for word in model_sg.wv.index_to_key:
        vector = model_sg.wv[word]
        print(f"{word}\t\t{vector}")

    print(f"\nWord2Vec (CBOW) Vectors with context window size = {context_size}:")
    model_cbow = train_word2vec_model(sentences, context_size, vector_size, sg=0)

    print("Word\t\tVector")
    for word in model_cbow.wv.index_to_key:
        vector = model_cbow.wv[word]
        print(f"{word}\t\t{vector}")

    print("------------------------")

Word2Vec (Skip-Gram) Vectors with context window size = 1:


Word		Vector
 		[-0.12376583 -0.04214299 -0.11454055 -0.01921923 -0.11364422 -0.23713072
  0.22867338  0.434367   -0.00631387 -0.56336     0.04938313 -0.30264318
  0.01359064  0.10035933  0.04786195  0.08331048  0.1994018   0.12580699
 -0.24938782 -0.28917935  0.0661598   0.21847151  0.10654852  0.1524928
 -0.17536065  0.00148325 -0.02853117  0.22429153 -0.3315946   0.22226495
  0.04157561  0.05615977  0.0312645  -0.10201073 -0.31346104  0.07694954
  0.1095252  -0.07971285 -0.09981567  0.0351567   0.06580926  0.03890392
 -0.03466644  0.28753325 -0.03111562 -0.14811938  0.01921378 -0.05353796
  0.17736943  0.0470863 ]
e		[-0.18112546  0.06577164 -0.16207814 -0.028069    0.13979088 -0.05437949
  0.08360896 -0.0394332  -0.26335895 -0.01212086  0.05375381  0.13999034
 -0.05951742 -0.09142236 -0.13300231  0.11961641  0.16494402  0.3847725
 -0.40492707 -0.01334396  0.03977606  0.08075628  0.26774997  0.09342119
 -0.00214226  0.06528804 -0.06157801  0.01386101 -0.3088798   0.05654854
  0.1553

Word		Vector
 		[-0.03068661 -0.09465004  0.27986142 -0.12730777  0.17649351  0.06913681
  0.06513873  0.08566421 -0.10695943 -0.26935524  0.00658547 -0.03748018
 -0.11632763  0.14991331  0.18759719  0.23260863  0.51967484  0.7045733
 -0.2334541  -0.14163764 -0.23476282 -0.13690288 -0.30009916  0.08570118
  0.17767906 -0.02193902  0.2682488   0.33208036 -0.17200796  0.35204133
  0.13097799  0.09134261  0.03450529 -0.4403707  -0.05966165  0.17234315
  0.37241465  0.17607036  0.07433017  0.45191073  0.51422685 -0.02046528
 -0.36659503  0.3226821   0.1779123   0.15913983 -0.34947267 -0.20330425
 -0.04562987 -0.06476699]
e		[-0.38840052 -0.09722094 -0.14582814 -0.12929448 -0.09038049  0.17856495
 -0.14065568 -0.6798344  -0.3324463   0.49026528  0.01168079  0.19718185
  0.21076874 -0.27588135 -0.04309884  0.26545495  0.01859758  0.20508234
 -0.33639085  0.1166392  -0.34344468 -0.01101853  0.3832144  -0.13019417
 -0.14446661  0.3863932   0.25950927 -0.1415391  -0.10732254  0.14265628
  0.591

Word		Vector
 		[-0.12293404 -0.09764976  0.13293868  0.11196119 -0.13406023 -0.23043317
  0.1370339   0.20126759 -0.11120888 -0.4567755  -0.10957977 -0.20071442
  0.05579802 -0.08526881  0.01238154  0.16891271  0.09434766  0.02145299
 -0.22066905 -0.0026789   0.12984546  0.07632604  0.4686309  -0.07219396
 -0.11393952  0.02722478 -0.21114294  0.09932285 -0.07700939  0.10680804
  0.02533181 -0.03034795  0.21175848 -0.06069995  0.13480029 -0.12453485
  0.15541562 -0.20387685 -0.02251509  0.19938166 -0.05735178 -0.17693073
 -0.1489736   0.06422628 -0.05875463 -0.2192692   0.08545773  0.10169704
 -0.00943383  0.01278721]
e		[-0.11176183 -0.05848807  0.07399675  0.12888347 -0.05132499 -0.14393808
 -0.0007276  -0.11460961 -0.056442   -0.05734761  0.04257831 -0.10987782
  0.17426492 -0.19908647  0.06955331  0.03476401  0.14965196  0.20298961
 -0.40087172  0.00984606 -0.06263708 -0.00627465  0.21744062  0.16889286
  0.09822941  0.04348435  0.0399506   0.17861086 -0.33278567  0.03136908
  0.04

Word		Vector
 		[-0.28154194 -0.21198933 -0.00540592 -0.28462452 -0.08348377 -0.08182792
 -0.01605504  0.16286898  0.5240778  -0.1782847   0.2782316   0.77723116
 -0.01601708 -0.03840789  0.13109112 -0.22043532  0.4561403   0.30063677
 -0.16074462  0.26070026  0.1427225  -0.35645372 -0.02342335  0.2292477
  0.20005196  0.08952843  0.27587312  0.38392374  0.00731449  0.39833993
  0.09582184  0.02913984  0.10049078 -0.13929287  0.27347866  0.20705175
  0.7781849   0.07855445 -0.10138419  0.50394213  0.32933     0.0835026
 -0.17001116 -0.1179075   0.11750792  0.04179762  0.16017044  0.00247845
 -0.10958403  0.34241918]
e		[-0.46525538 -0.05107431 -0.7271416   0.03062425  0.04476666  0.27489182
 -0.01524424 -0.4225412  -0.21144783  0.27733037 -0.14411703  0.12718362
 -0.23787598 -0.21140772  0.41888982  0.52761495 -0.10188685  0.11246626
 -0.4653377  -0.05729911 -0.26692265 -0.04738794 -0.08453408 -0.4273589
  0.38713458  0.16652776 -0.277633    0.43005824 -0.46535355 -0.11636452
  0.42561

Word		Vector
 		[ 0.01703177 -0.09920216  0.00858493  0.07855362  0.04567422 -0.13012594
  0.03357903  0.18228658 -0.04360858 -0.15558931 -0.0990076  -0.07658112
  0.17978722 -0.02087034  0.03774375  0.21450302  0.1555364   0.19189696
 -0.2889848  -0.20124874  0.13249904  0.0079683   0.33358657 -0.0878346
  0.11762251  0.02685517 -0.3242039  -0.08425539 -0.04089186 -0.06424485
  0.02783869  0.05424367  0.01680515 -0.04781545  0.05471064  0.14571483
  0.03428385 -0.05405585  0.00049745  0.04308082 -0.04271154 -0.2874753
 -0.07106509 -0.11487577  0.18405652  0.0733882  -0.00067675 -0.1527354
  0.29855114 -0.12300484]
e		[-0.20903724  0.03002428 -0.07059243 -0.00284687  0.11724343  0.058973
 -0.02719703  0.07983666  0.04524347  0.06002213  0.15754178 -0.09228513
  0.0549036  -0.07401203  0.08359998  0.06544285  0.14469706  0.1712522
 -0.1628911  -0.07014675 -0.01455734  0.01824892  0.10462366 -0.01632604
  0.04873665  0.06133589 -0.0050831   0.04028994 -0.15902328 -0.19177777
  0.13682786

Word		Vector
 		[-0.08593781 -0.14730579  0.10160777 -0.15834633 -0.34610274 -0.298316
 -0.24605174  0.03783828 -0.03057492 -0.2571755   0.31763437  0.23357196
 -0.04958284 -0.03202406  0.53168184  0.21634139  0.51146525  0.2881371
 -0.45213342  0.31552976 -0.15115288  0.08294405 -0.10611152 -0.17509827
  0.15116337  0.32785356  0.10498452  0.47943467 -0.00100998  0.8130217
 -0.08591039 -0.17854254 -0.11869292 -0.03514864  0.3621972   0.28566006
 -0.21032195  0.06376082  0.17929575  0.39132115  0.25034088  0.6617973
  0.1494608  -0.06817272 -0.04413078  0.23797409 -0.39136547  0.25652227
  0.30818456  0.08859646]
e		[-0.12863578 -0.11162135 -0.28964683 -0.10336312 -0.12798826 -0.15919307
  0.43731052 -0.6754104  -0.12157182  0.17797136  0.23716196  0.4921858
 -0.11167172 -0.12835847 -0.54612166  0.22151935 -0.19886819  0.6629905
 -0.2571149  -0.48164278 -0.00748163 -0.18332799  0.0851123  -0.2678173
 -0.0370783   0.72430795  0.24649255  0.23405348 -0.11772167 -0.40311274
 -0.05481226  

Word		Vector
 		[ 0.05901723 -0.05389757 -0.09241296  0.17549966  0.16095558 -0.0471079
  0.12309539  0.09576729 -0.04738633 -0.3147743   0.2771091  -0.15718561
  0.16745548 -0.14796786  0.04869439  0.06732379  0.01436027  0.18312727
 -0.30285934  0.0652108   0.11031357  0.19968906  0.1735064  -0.0213923
  0.11985308 -0.00562822 -0.08468423  0.13609734 -0.07806533 -0.14605224
  0.11467787 -0.116342   -0.25640166  0.04453945 -0.12241618  0.05788973
 -0.00325651 -0.0664839  -0.07921488 -0.00207331  0.11907214  0.04889106
  0.02505416 -0.00656174  0.03778602  0.0585962   0.02312207 -0.15396
 -0.15741403 -0.10539267]
e		[-0.03432079 -0.02515876 -0.18803678  0.11450548  0.01346922 -0.21345353
  0.03161675 -0.07220773 -0.02218367 -0.17145528  0.33430833 -0.01258596
 -0.04795872 -0.2064842   0.06179365 -0.03730107  0.08090363  0.06349511
 -0.19253911 -0.01749084  0.01510812  0.0857146   0.09292338 -0.06685189
  0.12203313  0.04124903 -0.00383193  0.17802148 -0.17091282  0.02996886
  0.1361304

Word		Vector
 		[ 0.3251143  -0.25064123  0.36213502 -0.43211633  0.03671788 -0.3708965
 -0.24730256  0.11154949  0.46020436 -0.53941065  0.62924576  0.5564313
  0.07496282 -0.01899807 -0.3268613  -0.625122    0.6960571  -0.09110707
  0.10052004 -0.22887711  0.1794673  -0.11936924 -0.00870845  0.14537816
  0.0946658   0.3001577  -0.0635022   0.12900825 -0.01447887  0.32859176
 -0.15372767 -0.11795515  0.09290583  0.05895375 -0.05063477 -0.13242579
  0.04562484 -0.00819181  0.3977851   0.15565777  0.4202563   0.04567094
  0.03545579 -0.13708125  0.19492161  0.01794446  0.62740594  0.05536039
  0.40631953  0.74062824]
e		[-0.5061682  -0.02148205  0.28103268  0.14750372  0.3236986   0.09411545
  0.31796896  0.05652115  0.03128069 -0.0860772   0.04391448 -0.5129913
  0.26660565  0.36315185  0.28423572  0.14321941 -0.19568953  0.41136795
 -0.6137476   0.27643025 -0.21776377 -0.07125255  0.4560883  -0.2779949
  0.26202565 -0.17178501 -0.2460809   0.5278356  -0.7255776  -0.40452427
  0.796161

Word		Vector
 		[-0.034281   -0.01142433 -0.04636988  0.0938819  -0.05074196  0.00321171
 -0.22825587  0.04918161 -0.03692305 -0.23104128  0.20167007  0.04363573
  0.02305413 -0.00250225  0.19518915  0.05642497 -0.05440515  0.19696014
 -0.22301093  0.17467587  0.00561398  0.07597653  0.27011153  0.12557822
 -0.06268406  0.07329164 -0.03873662  0.10421831 -0.04449461  0.24373226
  0.21744929 -0.11548065 -0.2006349  -0.12411548 -0.17924221 -0.014972
  0.08235619 -0.10232924 -0.00724413  0.03542895 -0.08360146 -0.10625504
 -0.11530337 -0.11665487 -0.04802986  0.12730937  0.02986834 -0.04960136
 -0.00125004 -0.29438126]
e		[-0.09492399 -0.06478114  0.01769386  0.06447315  0.11923565  0.03531994
  0.09292697  0.01187587  0.09546997 -0.02822131  0.14635456 -0.01536108
  0.06050745 -0.01015405  0.1049552  -0.14865007 -0.02841391  0.13613592
 -0.16702782  0.1901005   0.11305973  0.00624523  0.16539405 -0.07419194
 -0.08257788  0.04014349 -0.068499    0.04183679 -0.07296126  0.14343211
  0.1338

In [17]:
#Displaying the embeddings and interpreting the results for predicting similar
#words to 'fantastic'as  target word using the results of embeddings on different window sizes
#used in above task

import pandas as pd
from gensim.models import Word2Vec
from nltk.tokenize import word_tokenize

# Create a list of sentences
sentences = df['Cleaned_Text'].apply(lambda x: word_tokenize(x.lower())).tolist()

# Define target word for similarity comparison
target_word = 'fantastic'  # Replace with any target word you want to explore

for context_size in context_sizes:
    print(f"Word2Vec Vectors with context window size = {context_size}:")
    model = train_word2vec_model(sentences, context_size, vector_size, sg=1)

    print(f"Similar words to '{target_word}':")
    similar_words = model.wv.most_similar(target_word)
    for word, similarity in similar_words:
        print(f"{word}: {similarity}")

    print("------------------------")

Word2Vec Vectors with context window size = 1:
Similar words to 'fantastic':
awesome: 0.9579513669013977
fabulous: 0.9441646337509155
amazing: 0.9345090985298157
delightful: 0.9275104999542236
kallysten: 0.9072866439819336
terrific: 0.8912346959114075
wonderful: 0.8735024333000183
fascinating: 0.8733315467834473
joy: 0.8687556982040405
disappoint: 0.8675398826599121
------------------------
Word2Vec Vectors with context window size = 2:
Similar words to 'fantastic':
awesome: 0.9283010363578796
amazing: 0.9267646074295044
delightful: 0.8981768488883972
wonderful: 0.8854948878288269
fabulous: 0.8826448917388916
kallysten: 0.8729106187820435
captivating: 0.8656917810440063
delivers: 0.8441348671913147
disappoint: 0.8439163565635681
lovely: 0.8395051956176758
------------------------
Word2Vec Vectors with context window size = 3:
Similar words to 'fantastic':
awesome: 0.9098877906799316
amazing: 0.8979839086532593
fabulous: 0.8857145309448242
disappoint: 0.8845046758651733
stephanie: 0.881

Context Window Size 1:

The word 'fantastic' is highly similar to words such as 'awesome', 'fabulous', 'amazing', and 'delightful'.
The similarity scores for these words are quite high, ranging from 0.9 to 0.95.


Context Window Size 2:

Similar to the previous case, the word 'fantastic' is still strongly associated with words like 'awesome', 'amazing', and 'delightful'.
The similarity scores for these words remain consistently high, with values ranging from 0.88 to 0.93.


Context Window Size 3:

The similarity scores for 'fantastic' and related words remain relatively high, although slightly lower compared to the previous cases.
Words such as 'awesome', 'amazing', 'fabulous', and 'wonderful' continue to exhibit strong associations.


Context Window Size 4:

The similarity scores between 'fantastic' and other words remain reasonably high, with 'amazing', 'awesome', and 'wonderful' being the most similar.
Some new words like 'mallery' and 'andre' also appear with high similarity scores.


Context Window Size 5:

The similarity scores for 'fantastic' and related words are still relatively high, although slightly lower compared to previous cases.
Words like 'amazing', 'fabulous', and 'mckade' are still considered highly similar to 'fantastic'.
'vaughn', 'bdb' (possibly referring to Black Dagger Brotherhood), and 'd'arc' (possibly referring to D'Arc) also show up as similar words.


In conclusion, regardless of the context window size used, the word 'fantastic' consistently exhibits strong associations with positive words like 'awesome', 'amazing', 'fabulous', and 'wonderful'. The context window size affects the presence of additional words with high similarity scores, introducing some variations in the results. However, the overall pattern of strong similarity between 'fantastic' and positive descriptors remains consistent across different context window sizes.